src/daft-local-execution/src/intermediate_ops/hash_join_probe.rs

```diff
@@ -129,7 +296,32 @@ impl IntermediateOperator for HashJoinProbeOperator {
                     .downcast_mut::<HashJoinProbeState>()
                     .expect("HashJoinProbeOperator state should be HashJoinProbeState");
                 let input = input.as_data();
-                let out = state.probe(input, &self.right_on, &self.pruned_right_side_columns)?;
+                let out = match self.join_type {
+                    JoinType::Inner => state.probe_inner(
+                        input,
+                        &self.probe_on,
+                        &self.common_join_keys,
+                        &self.left_non_join_columns,
+                        &self.right_non_join_columns,
+                        self.build_on_left,
+                    ),
+                    JoinType::Left | JoinType::Right => state.probe_left_right(
+                        input,
+                        &self.probe_on,
+                        &self.common_join_keys,
+                        &self.left_non_join_columns,
+                        &self.right_non_join_columns,
+                        self.join_type == JoinType::Left,
+                    ),
+                    JoinType::Semi | JoinType::Anti => state.probe_anti_semi(
+                        input,
+                        &self.probe_on,
+                        self.join_type == JoinType::Semi,
+                    ),
+                    JoinType::Outer => {
+                        unimplemented!("Outer join is not yet implemented")
+                    }
```

Comment:
Not important but why do we have the actual probing logic inside of the state instead of the operator?

Reply:
Not much of a compelling reason, I think it's safe to move the logic to operator side and keep state logic free.